platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/linking/ConsensusLinker.java

```diff
@@ -0,0 +1,136 @@
+import java.time.Instant;
+
+/**
+ * Links events to their parents. Expects events to be provided in topological order.
+ * <p>
+ * This implementation adds extra metrics to the base class.
+ */
+public class ConsensusLinker extends InOrderLinker {
+
+    private final LongAccumulator missingParentAccumulator;
+    private final LongAccumulator generationMismatchAccumulator;
+    private final LongAccumulator birthRoundMismatchAccumulator;
+    private final LongAccumulator timeCreatedMismatchAccumulator;
+
+    private final StaleMetrics staleMetrics;
+
+    /**
+     * Constructor
+     *
+     * @param platformContext    the platform context
+     */
+    public ConsensusLinker(@NonNull final PlatformContext platformContext, @NonNull final NodeId selfId) {
+        super(platformContext);
+
+        missingParentAccumulator = platformContext
+                .getMetrics()
+                .getOrCreate(new LongAccumulator.Config(PLATFORM_CATEGORY, "missingParents")
+                        .withDescription("Parent child relationships where a parent was missing"));
+        generationMismatchAccumulator = platformContext
+                .getMetrics()
+                .getOrCreate(
+                        new LongAccumulator.Config(PLATFORM_CATEGORY, "parentGenerationMismatch")
+                                .withDescription(
+                                        "Parent child relationships where claimed parent generation did not match actual parent generation"));
+        birthRoundMismatchAccumulator = platformContext
+                .getMetrics()
+                .getOrCreate(
+                        new LongAccumulator.Config(PLATFORM_CATEGORY, "parentBirthRoundMismatch")
+                                .withDescription(
+                                        "Parent child relationships where claimed parent birth round did not match actual parent birth round"));
+        timeCreatedMismatchAccumulator = platformContext
+                .getMetrics()
+                .getOrCreate(
+                        new LongAccumulator.Config(PLATFORM_CATEGORY, "timeCreatedMismatch")
+                                .withDescription(
+                                        "Parent child relationships where child time created wasn't strictly after parent time created"));
+
+        staleMetrics = new StaleMetrics(platformContext, selfId);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void eventHasBecomeAncient(@NonNull final EventImpl event) {
+        if (!event.isConsensus()) {
+            staleMetrics.staleEvent(event);
+        }
+        event.clear();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void childHasMissingParent(
+            @NonNull final GossipEvent child, @NonNull final EventDescriptor parentDescriptor) {
+        super.childHasMissingParent(child, parentDescriptor);
+        missingParentAccumulator.update(1);
```

Comment:
do we expect there to be missing parents at this point? I thought the orphan buffer took care of that

Reply:
I agree, this should be impossible. Austin originally wrote the logic to catch this case -- I think the purpose was primarily as a sanity check.

Reply:
but why have a metric for that?

Reply:
Because we rate limit the log, we can't use the logs alone to determine how often it happens (if it happens).